### PR TITLE
eth: separate common and full node-specific API and backend service

### PR DESCRIPTION
**Pull request by zsfelfoldi**
This PR creates the ethapi package that implements the basic Ethereum API functions using an ethapi.Backend interface that is implemented in package eth (for full nodes) and will be also implemented in package les (for light nodes).
Full node-specific API functions are still implemented in the eth package (other APIs are also in their respective packages). If we want to change this and move everything to ethapi, that can be done in another PR (not sure if it is a practical idea though since then we would have to create interfaces for every service).
AccountManager is registered as a service now so that it can be easily accessed wherever needed. Not sure if it is the nicest solution but it was a simple one. If we're planning to get rid of it anyways then might do it for now.
The actual service-related code that can be shared between full and light nodes was mostly in the initialization. I've created three exported functions (CreateDBs, SetupGenesisBlock, CreatePoW) that can be used by the light service constructor too.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Fri Jan 29 03:14:49 UTC 2016_

**Comment by karalabe** [comment]
The RPC has some serious issues in it, some requiring suite some modifications in the eth API. I would wait a bit for the API to become stable again before doing any king of split. :(

**Comment by fjl** [comment]
Sorry for the delay, I will review this next week.

**Comment by zsfelfoldi** [comment]
@fjl thanks
@obscuren please also review when you have the time, I think your opinion is needed on this one.

**Comment by zsfelfoldi** [comment]
@karalabe PTAL at the abi binding

File: accounts/abi/bind/backends/remote.go [hunk 1]
```diff
@@ -98,9 +98,19 @@ func (b *rpcBackend) request(method string, params []interface{}) (json.RawMessa
 	if err := b.client.Send(req); err != nil {
 		return nil, err
 	}
+	//TODO: cancel request if needed after RPC cancel is implemented
```
Comment (zsfelfoldi) [inline accounts/abi/bind/backends/remote.go]: https://github.com/ethereum/go-ethereum/pull/2541 implements this

File: accounts/abi/bind/backends/remote.go [hunk 2]
```diff
@@ -80,27 +81,53 @@ type failure struct {
...
+	case err := <-errc:
+		if err != nil {
+			return nil, err
+		}
+	case <-ctx.Done():
+		// cancel RPC request
+		req := &request{
+			JSONRPC: "2.0",
+			ID:      int(atomic.AddUint32(&b.autoid, 1)),
+			Method:  "rpc_cancel",
```
Comment (zsfelfoldi) [inline accounts/abi/bind/backends/remote.go]: https://github.com/ethereum/go-ethereum/pull/2541 implements rpc_cancel

**Comment by fjl** [comment]
Summary of the discussion in the review call:

* ethapi will move to internal/ethapi. The purpose of package ethapi is to provide RPC handlers for both   light and full Ethereum nodes. The respective packages (eth, light) instantiate these handlers on top of their backend. go-ethereum library users should not use the RPC handlers as API.
* solc compiler can be removed from ethapi.Backend
* naming of eth.FullNodeService can be improved. @karalabe will improve names in a follow-up PR.
* AccountManager is not a Service and we will find a nice integration point for it. @fjl will clean up AccountManager integration in a follow-up PR
* RPC request cancellation should not be mixed with this PR. The new RPC client will provide cancellation through it's `CallContext` method.

**Comment by zsfelfoldi** [comment]
- fix the license in eth/api.go
- debug API functions could be moved to the common API (I'll see if it's easy or I leave it to a subsequent PR)

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -51,21 +51,32 @@ const defaultGas = uint64(90000)
...
 	b Backend
+	solcPath *string
+	solc     **compiler.Solidity
 }
 
 // NewPublicEthereumAPI creates a new Etheruem protocol API.
-func NewPublicEthereumAPI(b Backend) *PublicEthereumAPI {
-	return &PublicEthereumAPI{b}
+func NewPublicEthereumAPI(b Backend, solcPath *string, solc **compiler.Solidity) *PublicEthereumAPI {
+	return &PublicEthereumAPI{b, solcPath, solc}
```
Comment (fjl) [inline internal/ethapi/api.go]: Why are you passing in the instance as a double pointer? 
I think the only thing that needs to be passed here is the path as a plain string.
If the string is "", compiler.New will use the default path.

Comment (zsfelfoldi) [inline internal/ethapi/api.go]: because both PublicEthereumAPI and PrivateAdminAPI need to access/modify these.

Comment (fjl) [inline internal/ethapi/api.go]: I see why you did it like this now. The issue is that the admin API wants to update the path and the ethereum API wants to use it. I will clean this up later, too, so the *compiler.Solidity handles this internally by providing SetPath.

**Comment by zsfelfoldi** [comment]
@karalabe @fjl I think the issues are addressed. I realized the print block debug functions were already in ethapi, just left some duplicates in the full node api by accident. The trace functions are not moved to ethapi yet, it can be done but that will need a bit of work. Will do those in a later PR after I rebased the light client on this one so that I can actually see those functions work in a light client environment before submitting the changes.

**Comment by fjl** [comment]
👍

**Comment by fjl** [comment]
Please gofmt.

**Comment by fjl** [comment]
Please use LGPL license headers in internal/ethapi/*.go.

**Comment by zsfelfoldi** [comment]
done

**Comment by zsfelfoldi** [comment]
will squash commits after it gets a thumbs up from @karalabe too :)

File: accounts/abi/bind/backends/remote.go [hunk 4]
```diff
@@ -80,27 +81,41 @@ type failure struct {
...
 		Method:  method,
 		Params:  params,
 	}
 	if err := b.client.Send(req); err != nil {
 		return nil, err
 	}
 	res := new(response)
-	if err := b.client.Recv(res); err != nil {
-		return nil, err
+	errc := make(chan error)
```
Comment (karalabe) [inline accounts/abi/bind/backends/remote.go]: Make this a buffered channel with a buffer of 1. Otherwise if the context is cancelled noone will read from the error channel and the goroutine will be leaked.

File: accounts/abi/bind/base.go [hunk 5]
```diff
@@ -35,6 +36,8 @@ type SignerFn func(common.Address, *types.Transaction) (*types.Transaction, erro
 // CallOpts is the collection of options to fine tune a contract call request.
 type CallOpts struct {
 	Pending bool // Whether to operate on the pending state or the last known one
+
+	Ctx context.Context // unlike usual practice, this context can be left uninitialized (nil)
```
Comment (karalabe) [inline accounts/abi/bind/base.go]: Please rename this field to Context. I'd like to keep the API consistent. Also please change the comment to `Network context to support cancellation and timeouts (nil = no timeout)`.

File: accounts/abi/bind/base.go [hunk 6]
```diff
@@ -47,6 +50,8 @@ type TransactOpts struct {
 	Value    *big.Int // Funds to transfer along along the transaction (nil = 0 = no funds)
 	GasPrice *big.Int // Gas price to use for the transaction execution (nil = gas price oracle)
 	GasLimit *big.Int // Gas limit to set for the transaction execution (nil = estimate + 10%)
+
+	Ctx context.Context // unlike usual practice, this context can be left uninitialized (nil)
```
Comment (karalabe) [inline accounts/abi/bind/base.go]: Same as for CallOpts

File: core/vm/environment.go [hunk 7]
```diff
@@ -73,6 +73,8 @@ type Environment interface {
 	DelegateCall(me ContractRef, addr common.Address, data []byte, gas, price *big.Int) ([]byte, error)
 	// Create a new contract
 	Create(me ContractRef, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error)
+
+	StructLogs() []StructLog
```
Comment (karalabe) [inline core/vm/environment.go]: Why is this here? Perhaps a merge error?

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 73e27bd865c07003c851ef37a55448d35294e20b
  * Commits must be prefixed with the package(s) they modify

 * Commit: d1e0fd4d5f80504f176aed1e206ba48648b7cf37
  * Commits must be prefixed with the package(s) they modify

 * Commit: c390528d66c31550295adbcaf8dc9793e183b35c
  * Commits must be prefixed with the package(s) they modify

 * Commit: 15f40a56501f3ca37a7040e53eebcec599561ed7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9252d63c9e7e930136b202812fca65f096d60a48
  * Commits must be prefixed with the package(s) they modify

 * Commit: 13c4d240d43cfc5b561710607f17c46de4715156
  * Commits must be prefixed with the package(s) they modify

 * Commit: be65a31edc9e2f9f90ff8be326e8431ed50e609e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1e150536ef109370ab9d1a1d224b509aba0340d7
  * Commits must be prefixed with the package(s) they modify

 * Commit: d3829caee270bf98b634915d41e172b34a9f17fb
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: core/vm/environment.go (hunk 7 above)
Comment (zsfelfoldi) [inline core/vm/environment.go]: No, the reason is that ethapi.Backend uses the vm.Environment interface instead of *core.VMEnv (needs to be replaceable, there is a light implementation for VMEnv) and TraceCall needs to access StructLogs.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 73e27bd865c07003c851ef37a55448d35294e20b
  * Commits must be prefixed with the package(s) they modify

 * Commit: d1e0fd4d5f80504f176aed1e206ba48648b7cf37
  * Commits must be prefixed with the package(s) they modify

 * Commit: c390528d66c31550295adbcaf8dc9793e183b35c
  * Commits must be prefixed with the package(s) they modify

 * Commit: 15f40a56501f3ca37a7040e53eebcec599561ed7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9252d63c9e7e930136b202812fca65f096d60a48
  * Commits must be prefixed with the package(s) they modify

 * Commit: 13c4d240d43cfc5b561710607f17c46de4715156
  * Commits must be prefixed with the package(s) they modify

 * Commit: be65a31edc9e2f9f90ff8be326e8431ed50e609e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1e150536ef109370ab9d1a1d224b509aba0340d7
  * Commits must be prefixed with the package(s) they modify

 * Commit: d3829caee270bf98b634915d41e172b34a9f17fb
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9fe07656ad187dc26d822d06518f3006dc01471b
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: internal/ethapi/api.go [hunk 8]
```diff
@@ -615,9 +608,9 @@ func (s *PublicBlockChainAPI) doCall(ctx context.Context, args CallArgs, blockNr
 			return "0x", gas, err
 		}
 		return common.ToHex(res), gas, err
+	} else {
+		return "0x", common.Big0, err
 	}
```
Comment (karalabe) [inline internal/ethapi/api.go]: There's no need for this else branch here. The if branch already returns. Please revert to the original code.

File: internal/ethapi/api.go [hunk 9]
```diff
@@ -734,9 +727,9 @@ func (s *PublicBlockChainAPI) TraceCall(ctx context.Context, args CallArgs, bloc
 			ReturnValue: fmt.Sprintf("%x", ret),
 			StructLogs:  FormatLogs(vmenv.StructLogs()),
 		}, nil
+	} else {
+		return nil, err
 	}
```
Comment (karalabe) [inline internal/ethapi/api.go]: Similarly, there's no need for an else branch here, please drop and revert to the original code.

File: internal/ethapi/api.go (hunk 8 above)
Comment (karalabe) [inline internal/ethapi/api.go]: Ah I see, you need it for the error. Then Please do the error handling first and then the long code (outside of the if still!)

File: eth/api_backend.go [hunk 10]
```diff
@@ -102,12 +97,8 @@ func (b *EthApiBackend) GetTd(blockHash common.Hash) *big.Int {
...
 }
 
-func (b *EthApiBackend) GetVMEnv(ctx context.Context, msg core.Message, header *types.Header) (vm.Environment, func() error, error) {
-	stateDb, err := state.New(header.Root, b.eth.chainDb)
-	if err != nil {
-		return nil, nil, err
-	}
-	stateDb = stateDb.Copy()
+func (b *EthApiBackend) GetVMEnv(ctx context.Context, msg core.Message, state ethapi.State, header *types.Header) (vm.Environment, func() error, error) {
+	stateDb := state.(EthApiState).state.Copy()
```
Comment (karalabe) [inline eth/api_backend.go]: Why do you copy here? AFAIK the state should already be copied when provided to you. Am I mistaking?

Comment (zsfelfoldi) [inline eth/api_backend.go]: Previously it was copied by the caller but now we use ethapi.State interfaces that don't have a Copy function (or any modify function, they're read-only views). So I thought it is easier and maybe also more logical to copy here so that we keep the immutable property of ethapi.State.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 73e27bd865c07003c851ef37a55448d35294e20b
  * Commits must be prefixed with the package(s) they modify

 * Commit: d1e0fd4d5f80504f176aed1e206ba48648b7cf37
  * Commits must be prefixed with the package(s) they modify

 * Commit: c390528d66c31550295adbcaf8dc9793e183b35c
  * Commits must be prefixed with the package(s) they modify

 * Commit: 15f40a56501f3ca37a7040e53eebcec599561ed7
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9252d63c9e7e930136b202812fca65f096d60a48
  * Commits must be prefixed with the package(s) they modify

 * Commit: 13c4d240d43cfc5b561710607f17c46de4715156
  * Commits must be prefixed with the package(s) they modify

 * Commit: be65a31edc9e2f9f90ff8be326e8431ed50e609e
  * Commits must be prefixed with the package(s) they modify

 * Commit: 1e150536ef109370ab9d1a1d224b509aba0340d7
  * Commits must be prefixed with the package(s) they modify

 * Commit: d3829caee270bf98b634915d41e172b34a9f17fb
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9fe07656ad187dc26d822d06518f3006dc01471b
  * Commits must be prefixed with the package(s) they modify

 * Commit: 3a8b3bc2f72125ae7996d00807027c26683b911a
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: internal/ethapi/api.go (hunk 8 above)
Comment (zsfelfoldi) [inline internal/ethapi/api.go]: done

**Comment by karalabe** [comment]
Let's get this in finally!
